### PR TITLE
8341407: C2: assert(main_limit == cl->limit() || get_ctrl(main_limit) == new_limit_ctrl) failed: wrong control for added limit

### PR DESCRIPTION
**Pull request by rwestrel**
That assert checks that during RC elimination, we have either:

- not updated the limit of the main loop

- or that the new limit is at the expected control

The assert fires because the limit was updated but is not at the
expected control. That happens because `new_limit_ctrl` is updated for
a test that it attempts to eliminate before it actually proceeds with
the elimination: if the test can't be eliminated, `new_limit_ctrl`
gets updated anyway.

While the assert could, maybe, be relaxed (it fires in this case but
nothing is going wrong), it's better, I think, to simply not uselessly
restrict the control of the limit.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8341407](https://bugs.openjdk.org/browse/JDK-8341407): C2: assert(main_limit == cl-&gt;limit() || get_ctrl(main_limit) == new_limit_ctrl) failed: wrong control for added limit (**Bug** - P3)

### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) Review applies to [ccb5c07c](https://git.openjdk.org/jdk/pull/21564/files/ccb5c07c3c7316e69a8a5f451d7dcda5c05342f1)
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21564/head:pull/21564` \
`$ git checkout pull/21564`

Update a local copy of the PR: \
`$ git checkout pull/21564` \
`$ git pull https://git.openjdk.org/jdk.git pull/21564/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21564`

View PR using the GUI difftool: \
`$ git pr show -t 21564`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21564.diff">https://git.openjdk.org/jdk/pull/21564.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21564#issuecomment-2419658505)

### CODE REVIEW

File: src/hotspot/share/opto/loopTransform.cpp [hunk 1]
```diff
@@ -3001,6 +3001,9 @@ void PhaseIdealLoop::do_range_check(IdealLoopTree *loop, Node_List &old_new) {
           continue;             // Unhandled case
         }
       }
+      // Only updated variable tracking control for new nodes if it's indeed a range check that can be eliminated (and
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/loopTransform.cpp:3004]:
```suggestion
      // Only update variable tracking control for new nodes if it's indeed a range check that can be eliminated (and
```

**Review by TobiHartmann — APPROVED** [review]
Looks good to me.

**Comment by rwestrel** [comment]
@TobiHartmann @chhagedorn thanks for the reviews

**Pull request closed without merging**